chore: let `Function::inlined` take a `should_inline_call` function

asterite opened the pull request:
# Description

## Problem

Part of #7148

## Summary



## Additional Context



## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

TomAFrench commented:
If we're marking this as resolving #7148, can you break out issues for all the extra inlining conditions it specifies?

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/inlining.rs:
```diff
@@ -68,8 +68,16 @@ impl Ssa {
-            let new_function =
-                function.inlined(&self, inline_no_predicates_functions, inline_infos);
+            let should_inline_call =
+                |_context: &PerFunctionContext, ssa: &Ssa, called_func_id: FunctionId| -> bool {
```
Can you remove the `_context` arg if this is forcing us to expose `PerFunctionContext` beyond the inlining pass?

asterite commented:
I changed it to be "part of".

Regarding the extra inlining conditions, are those to be done in `preprocess_fns.rs`?

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/inlining.rs:
```diff
@@ -81,47 +89,47 @@ impl Function {
+        // Do not inline self-recursive functions on the top level.
+        // Inlining a self-recursive function works when there is something to inline into
+        // by importing all the recursive blocks, but for the entry function there is no wrapper.
+        if called_func_id == self.id() {
+            return false;
+        }
```
Note that these checks should be enforced even if we're passing a different inlining condition in as the closure as theyre important safety checks. Can you pull them out of this function and ensure they're always enforced?

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/inlining.rs:
```diff
@@ -81,47 +89,47 @@ impl Function {
+                !inline_type.is_entry_point() && !preserve_function
+            }
+            RuntimeType::Brillig(_) => {
+                if self.runtime().is_acir() {
+                    // We never inline a brillig function into an ACIR function.
+                    return false;
```
ditto here and for inlining ACIR into brillig.

asterite replied:
> and for inlining ACIR into brillig

Is that a new check that should be added?

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/inlining.rs:
```diff
@@ -81,47 +89,47 @@ impl Function {
+    }
+
+    /// Generic function that determines whether a function should inline a call.
+    pub(super) fn should_inline_call(
```
We don't want to enshrine this as the only way of determining whether a function call should be inlined or not (quite the opposite) so I don't think it should be defined as a method on `Function`.

TomAFrench commented:
> at least the first two which seem simpler than the last one

This should be able to be calculated using `compute_times_called`.

TomAFrench commented:
> Regarding the extra inlining conditions, are those to be done in preprocess_fns.rs?

I don't think so as `preprocess_fns` is at risk of becoming a mega-pass which does too much. These early inlining passes can be done prior to this.

asterite commented:
I addressed all the comments (I think). To avoid a double lookup in `ssa.functions` I changed the callback to receive `&Function`, and now it doesn't need to take an `&Ssa` anymore (and for the future optimizations I think that'll also remain like that, as we can check what's in a function without going through `Ssa`). Let me know if that's okay, otherwise we can revert the last commit.

asterite commented on compiler/noirc_evaluator/src/ssa/opt/inlining.rs:
```diff
@@ -882,14 +872,46 @@ impl<'function> PerFunctionContext<'function> {
+                    return None;
+                }
+            }
+        }
```
Done! I couldn't easily check that the two runtime types match (because of `InlineType`) so I compared the bools of `is_acir()`.

asterite replied:
This change broke things. It seems we so inline acir functions into brillig so I reverted that last commit.

TomAFrench replied:
Huh, can you open an issue to investigate this as it's not immediately obvious why this should happen after the SSA gen split?

My money would be on some special-cased functions such as built-ins / blackbox functions not being duplicated in the same way that regular noir functions are handled.

asterite replied:
I just checked. One program that was failing was `array_sort`, which I think doesn't have built-ins or blackbox functions. Printing the conflicting functions, it was trying to inline an acir lambda into a brillig function. Maybe lambdas aren't separated?

asterite replied:
Here's a small program that reproduces the issue:

```noir
fn main() -> pub Field {
    foo(|x| x + 1)
}

fn foo(f: fn(Field) -> Field) -> Field {
    unsafe {
        bar(f)
    }
}

unconstrained fn bar(f: fn(Field) -> Field) -> Field {
    f(1)
}
```

The SSA is this:

```
acir(inline) fn main f0 {
  b0():
    v2 = call f2(f1) -> Field
    return v2
}
acir(inline) fn lambda f1 {
  b0(v0: Field):
    v2 = add v0, Field 1
    return v2
}
acir(inline) fn foo f2 {
  b0(v0: function):
    v2 = call f3(v0) -> Field
    return v2
}
brillig(inline) fn bar f3 {
  b0(v0: function):
    v2 = call v0(Field 1) -> Field
    return v2
}
```

What's the SSA pass that separates runtimes?

asterite commented:
There's clearly something wrong but I can't tell what. I'll continue debugging this tomorrow.

asterite commented:
It's working! The performance alter was likely due to things not properly inlining anymore.

TomAFrench commented on compiler/noirc_evaluator/src/ssa/opt/inlining.rs:
```diff
@@ -205,7 +185,7 @@ pub(super) struct InlineInfo {
     is_acir_entry_point: bool,
     is_recursive: bool,
-    should_inline: bool,
+    pub(super) should_inline: bool,
```
```suggestion
    should_inline: bool,
```

TomAFrench reviewed (approved):
LGTM
